File: Medium/19_Remove_Nth_Node_From_End_of_List/Solution.py
from typing import Optional
from devlibs.singly_linkedlist import ListNode, printSinglyLinkList


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next


class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        n1 = head
        n2 = head

        for i in range(n):
            n2 = n2.next
        
        if (n2 == None):
            head = head.next
            return head

        while (n2.next != None):
            n2 = n2.next
            n1 = n1.next
            
        n1.next = n1.next.next
        
        return head


if __name__ == "__main__":
    solution = Solution()

    # Test Case 1:
    head1 = ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5)))))
    result1 = solution.removeNthFromEnd(head1, 2)
    printSinglyLinkList(result1)
    
    # Test Case 2:
    head2 = ListNode(1)
    result2 = solution.removeNthFromEnd(head2, 1)
    printSinglyLinkList(result2)

    
    # Test Case 3:
    head3 = ListNode(1, ListNode(2))
    result3 = solution.removeNthFromEnd(head3, 2)
    printSinglyLinkList(result3)

    

File: Medium/19_Remove_Nth_Node_From_End_of_List/Solution.java



class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode n1 = head;
        ListNode n2 = head;

        for (int i = 0; i < n; i++) {
            n2 = n2.next;
        }

        if (n2 == null) {
            head = head.next;
            return head;
        }

        while (n2.next != null) {
            n2 = n2.next;
            n1 = n1.next;
        }

        n1.next = n1.next.next;

        return head;
    }

    public static void main(String[] args) {
        Solution solution = new Solution();

        // test cases 1
        ListNode head1 = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))));
        int n1_val = 2;
        ListNode result1 = solution.removeNthFromEnd(head1, n1_val);
        while (result1 != null) {
            System.out.print(result1.val + " ");
            result1 = result1.next;
        }
        System.out.println();

        // test cases 2
        ListNode head2 = new ListNode(1);
        int n2_val = 1;
        ListNode result2 = solution.removeNthFromEnd(head2, n2_val);
        if (result2 == null) System.out.print("[]");
        while (result2 != null) {
            System.out.print(result2.val + " ");
            result2 = result2.next;
        }
        System.out.println();

        // test cases 3
        ListNode head3 = new ListNode(1, new ListNode(2));
        int n3_val = 1;
        ListNode result3 = solution.removeNthFromEnd(head3, n3_val);
        while (result3 != null) {
            System.out.print(result3.val + " ");
            result3 = result3.next;
        }
        System.out.println();
        
    }
}

// Definition for singly-linked list.
class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}